README.md
# workflows-with-secure-device-access-client

`workflows-with-secure-device-access-client` is an Mbed OS application that communicates with the Pelion Device Management mobile application and executes Secure Device Access (SDA) authorized and authenticated workflows. For information about SDA, please see the [public SDA documentation](https://www.pelion.com/docs/device-management/current/device-management/secure-device-access.html).

The application implements a Bluetooth Low Energy (BLE) communication interface with Pelion Device Management mobile application and integrates [littlefs](https://github.com/ARMmbed/littlefs) to read and write files to and from the target device.

This document explains how to build `workflows-with-secure-device-access-client` and flash the binary onto your device:

- [Prerequisites](#prerequisites)
- [Building the application](#building-the-application)
- [Flashing the binary](#flashing-the-binary)
- [Debugging](#debugging)
- [Troubleshooting](#troubleshooting)

## Prerequisites

### Hardware requirements
- An [ST DISCO-L475VG-IOT01A](https://os.mbed.com/platforms/ST-Discovery-L475E-IOT01A/) board.
- A USB cable (USB Type A to Micro-AB).

### Working environment requirements
- A Device Management account with the secure device access feature and workflow management service activated.
- Operating systems we have tested: Ubuntu 18.04.

### Library dependencies
- mbed-os v5.15.1
- mbed-cloud-client v4.2.1

Mbed CLI installs these dependencies when you run the `mbed deploy` command as part of the [build steps](#build-steps).

### Tools and toolchains
- [Install Python 2.7](https://www.python.org/download/releases/2.7/).
- [Install Mbed CLI](https://os.mbed.com/docs/mbed-os/v5.15/quick-start/offline-with-mbed-cli.html).
- [GNU Arm GCC Compiler](https://www.pelion.com/docs/device-management/current/cloud-requirements/tool-requirements.html#notes-for-toolchains).
- [pyOCD](https://pypi.org/project/pyocd/).
- [Update udev rules](https://github.com/mbedmicro/pyOCD/tree/master/udev).
    To update, download the udev rules directory to your system, navigate to the `udev` directory and run:
    ```
    sudo cp *.rules /etc/udev/rules.d
    ```
- libusb
    ```
    apt-get install libusb-1.0-0-dev
    ```
After you install Mbed CLI and GNU Arm GCC Compiler, set the path to the Arm embedded toolchain:
```
mbed config -G GCC_ARM_PATH "~/<gcc-arm-none-eabi-6-xxx>/bin"
```

## Building the application

1. Clone the application and navigate to the folder:

    ```
    git clone https://github.com/armPelionEdge/workflows-with-secure-device-access-client
    cd workflows-with-secure-device-access-client
    ```

1. Provision device credentials:

    - If you are using [developer credentials](https://www.pelion.com/docs/device-management/current/sda/obtaining-a-bootstrap-certificate-and-trust-anchor.html#in-developer-mode):
        1. Overwrite the `mbed_cloud_dev_credentials.c` template file in the `workflows-with-secure-device-access-client` folder with the `mbed_cloud_dev_credentials.c` file that you [download from the Portal](https://www.pelion.com/docs/device-management/current/sda/obtaining-a-bootstrap-certificate-and-trust-anchor.html#generate-a-bootstrap-certificate).
        1. From the `workflows-with-secure-device-access-client` folder, overwrite the `mbed_cloud_trust_anchor_credentials.c` template file with your [trust anchor](https://www.pelion.com/docs/device-management/current/sda/obtaining-a-bootstrap-certificate-and-trust-anchor.html#obtain-a-trust-anchor):
            ```
            python create_trust_anchor_dev_cred.py -t "<trust anchor>"
            ```
            For example:
            ```
            python create_trust_anchor_dev_cred.py -t "-----BEGIN PUBLIC KEY-----MFkaEqYHKoZIzj0CAQYIKoZIzj0GAQcDHgAEbiRnZgdzoBpySFDPVPFp3J7yOmrOXJ09O5qVUMOD5knUjX7YbQBF0ueJWPz6tkTGbzORAwDzvRXYUA7vZpB+og==-----END PUBLIC KEY-----"
            ```
            Note: Be sure to remove any line breaks (\n or \r) in the trust anchor before running `create_trust_anchor_dev_cred.py`.

    - In production mode, provision and store your credentials on your device, as described in the [provisioning documentation](https://www.pelion.com/docs/device-management/current/provisioning-process/index.html).

1. Install library dependencies:
    ```
    mbed deploy
    ```
    This downloads all libraries required to build the application.

1. Navigate to the `mbed-os` folder in `workflows-with-secure-device-access-client` and apply Mbed OS patches:
    ```
    git am ../patches/*
    ```

1. Configure the BLE interface by setting values for the parameters defined in `ble-config.h`.

    Note: `BLE_PACKET_SIZE` should be 12 to 232 bytes. We recommend you use a large packet size to reduce the number of BLE packets, increase throughput and reduce power consumption.

    Example:

    ```C
    #define BLE_PACKET_SIZE             200
    #define Device_Local_Name           "DISCO"
    #define MANUFACTURER_NAME           "ARM"
    #define MODEL_NUM                   "0.1.1"
    #define HARDWARE_REVISION           "0.1.1"
    #define FIRMWARE_REVISION           "0.1.1"
    #define SOFTWARE_REVISION           "0.1.1"
    #define MOUNT_POINT                 "fs"

    const uint8_t  ServiceBaseUUID[LONG_UUID_LENGTH] = {
        0x6E, 0x40, 0x00, 0x00, 0xB5, 0xA3, 0xF3, 0x93,
        0xE0, 0xA9, 0xE5, 0x0E, 0x24, 0xDC, 0xCA, 0x9E,
    };

    const uint16_t ServiceShortUUID                 = 0x0001;
    const uint16_t CharacteristicShortUUID          = 0x0002;

    const uint8_t  ServiceUUID[LONG_UUID_LENGTH] = {
        0x6E, 0x40, (uint8_t)(ServiceShortUUID >> 8), (uint8_t)(ServiceShortUUID & 0xFF), 0xB5, 0xA3, 0xF3, 0x93,
        0xE0, 0xA9, 0xE5, 0x0E, 0x24, 0xDC, 0xCA, 0x9E,
    };

    const uint8_t  CharacteristicUUID[LONG_UUID_LENGTH] = {
        0x6E, 0x40, (uint8_t)(CharacteristicShortUUID >> 8), (uint8_t)(CharacteristicShortUUID & 0xFF), 0xB5, 0xA3, 0xF3, 0x93,
        0xE0, 0xA9, 0xE5, 0x0E, 0x24, 0xDC, 0xCA, 0x9E,
    };
    ```

1. Compile the application:
    ```
    mbed compile -t GCC_ARM -m DISCO_L475VG_IOT01A
    ```

## Flashing the binary

1. Connect the board to your computer using USB.
1. In the terminal, run:
    ```
    pyocd gdbserver
    ```
    Press Enter while holding the Reset button on the board.

1. In another terminal, navigate to the `BUILD` folder in `workflows-with-secure-device-access-client`, and run:
    ```
    arm-none-eabi-gdb
    ```
    This launches the GDB (GNU Debugger) terminal.

1. To connect to the pyOCD server, inside the GDB terminal, run:
    ```
    target remote:3333
    ```

1. Load the binary onto your board:
    ```
    load workflows-with-secure-device-access-client.hex
    ```
    The LED blinks rapidly. Wait for it to finish.

## Debugging

1. Connect the board to your computer using USB.
1. Use `mbedls` to detect and list Mbed Enabled devices connected to the computer.
1. Use PuTTY or any other serial console with 115200 baud 8-N-1 configuration to view the logs.

Note: To change the trace level of logs, open the `mbed_app.json` file in the root folder of the project and change the value of `MBED_TRACE_MAX_LEVEL` from `TRACE_LEVEL_ERROR` to `TRACE_LEVEL_INFO`. Compile the project and flash the new binary.

## Troubleshooting

Please see [GitHub issues](https://github.com/armPelionEdge/workflows-with-secure-device-access-client/issues) for solutions to common build errors.


CHANGELOG.md
# Changelog

## Release v1.0.0 - 2020-05-27

### Features
- Supported target [ST DISCO-L475VG-IOT01A](https://os.mbed.com/platforms/ST-Discovery-L475E-IOT01A/).
- Integrated littlefs to mange the files on the target device.
    - Supports filename of length 256.
    - Default mount point is `/fs`.
- Supported workflow task types `configure` and `read-data`.
- Works with v1.0.1 of PDM mobile app.
- Using Mbed OS version 5.15.
- Using Device Management Client version 4.2.1.
- Provide BLE communication interface with PDM mobile application.
- Provide options to configure BLE discover service uuid and service characteristics uuid. 
- Provide option to configure BLE device local name, manufacture name, model number,  hardware revision, firmware revision and software revision.
- Developed and implemented a simple low-overhead fragmentation  protocol to handle packets bigger than the BLE MTU size.
- Read and write max file size is 1kB.
- Execute workflows which are authenticated by SDA client part of PDMC.

blesda.h
/* mbed Microcontroller Library
 * Copyright (c) 2006-2013 ARM Limited
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

#ifndef _BLE_SDA_SERVICE_H_
    #define _BLE_SDA_SERVICE_H_
#include "ble/UUID.h"
#include "ble/BLE.h"
#include "ble/pal/Deprecated.h"
#include "sda_interface.h"
#include "protocoltranslator.h"

const uint8_t  UARTServiceBaseUUID[UUID::LENGTH_OF_LONG_UUID] = {
    0x6E, 0x40, 0x00, 0x00, 0xB5, 0xA3, 0xF3, 0x93,
    0xE0, 0xA9, 0xE5, 0x0E, 0x24, 0xDC, 0xCA, 0x9E,
};
const uint16_t UARTServiceShortUUID                 = 0x0001;
const uint16_t UARTServiceTXCharacteristicShortUUID = 0x0002;
const uint16_t UARTServiceRXCharacteristicShortUUID = 0x0003;
const uint8_t  UARTServiceUUID[UUID::LENGTH_OF_LONG_UUID] = {
    0x6E, 0x40, (uint8_t)(UARTServiceShortUUID >> 8), (uint8_t)(UARTServiceShortUUID & 0xFF), 0xB5, 0xA3, 0xF3, 0x93,
    0xE0, 0xA9, 0xE5, 0x0E, 0x24, 0xDC, 0xCA, 0x9E,
};
const uint8_t  UARTServiceUUID_reversed[UUID::LENGTH_OF_LONG_UUID] = {
    0x9E, 0xCA, 0xDC, 0x24, 0x0E, 0xE5, 0xA9, 0xE0,
    0x93, 0xF3, 0xA3, 0xB5, (uint8_t)(UARTServiceShortUUID & 0xFF), (uint8_t)(UARTServiceShortUUID >> 8), 0x40, 0x6E
};
const uint8_t  UARTServiceTXCharacteristicUUID[UUID::LENGTH_OF_LONG_UUID] = {
    0x6E, 0x40, (uint8_t)(UARTServiceTXCharacteristicShortUUID >> 8), (uint8_t)(UARTServiceTXCharacteristicShortUUID & 0xFF), 0xB5, 0xA3, 0xF3, 0x93,
    0xE0, 0xA9, 0xE5, 0x0E, 0x24, 0xDC, 0xCA, 0x9E,
};
const uint8_t  UARTServiceRXCharacteristicUUID[UUID::LENGTH_OF_LONG_UUID] = {
    0x6E, 0x40, (uint8_t)(UARTServiceRXCharacteristicShortUUID >> 8), (uint8_t)(UARTServiceRXCharacteristicShortUUID & 0xFF), 0xB5, 0xA3, 0xF3, 0x93,
    0xE0, 0xA9, 0xE5, 0x0E, 0x24, 0xDC, 0xCA, 0x9E,
};

class BLESDA {
public:
    /** Maximum length of data (in bytes) that the UART service module can transmit to the peer. */
    static const unsigned BLE_UART_SERVICE_MAX_DATA_LEN = 244;//(BLE_GATT_MTU_SIZE_DEFAULT - 3);

public:

    /**
    * @param _ble
    *               BLE object for the underlying controller.
    */
    BLESDA(BLE &_ble, char* endpointbuffer) :
        ble(_ble),
        receiveBuffer(),
        sendBuffer(),
        sendBufferIndex(0),
        numBytesReceived(0),
        receiveBufferIndex(0),
        _endpointBuffer(endpointbuffer),
        txCharacteristic(UARTServiceTXCharacteristicUUID, receiveBuffer, 1, BLE_UART_SERVICE_MAX_DATA_LEN,
                         GattCharacteristic::BLE_GATT_CHAR_PROPERTIES_WRITE | GattCharacteristic::BLE_GATT_CHAR_PROPERTIES_WRITE_WITHOUT_RESPONSE),
        rxCharacteristic(UARTServiceRXCharacteristicUUID, sendBuffer, 1, BLE_UART_SERVICE_MAX_DATA_LEN, GattCharacteristic::BLE_GATT_CHAR_PROPERTIES_NOTIFY)
        {
            GattCharacteristic *charTable[] = {&txCharacteristic, &rxCharacteristic};
            GattService SDAService(UARTServiceUUID, charTable, sizeof(charTable) / sizeof(GattCharacteristic *));
            ble.gattServer().addService(SDAService);
            ble.gattServer().onDataWritten(this, &BLESDA::onDataWritten);
        }

    /**
     * Note: TX and RX characteristics are to be interpreted from the viewpoint of the GATT client using this service.
     */
    uint16_t getTXCharacteristicHandle();
    /**
     * Note: TX and RX characteristics are to be interpreted from the viewpoint of the GATT client using this service.
     */
    uint16_t getRXCharacteristicHandle();
    /*Need UUID of the service so that we can advertise that in the advertisement */

    const uint8_t* getUUID(){
        return UARTServiceUUID;
    }
    /**
     * We attempt to collect bytes before pushing them to the UART RX
     * characteristic; writing to the RX characteristic then generates
     * notifications for the client. Updates made in quick succession to a
     * notification-generating characteristic result in data being buffered
     * in the Bluetooth stack as notifications are sent out. The stack has
     * its limits for this buffering - typically a small number under 10.
     * Collecting data into the sendBuffer buffer helps mitigate the rate of
     * updates. But we shouldn't buffer a large amount of data before updating
     * the characteristic, otherwise the client needs to turn around and make
     * a long read request; this is because notifications include only the first
     * 20 bytes of the updated data.
     *
     * @param  _buffer The received update.
     * @param  length Number of characters to be appended.
     * @return        Number of characters appended to the rxCharacteristic.
     */
    size_t write(const void *_buffer, size_t length);
    /**
     * Helper function to write out strings.
     * @param  str The received string.
     * @return     Number of characters appended to the rxCharacteristic.
     */
    size_t writeString(const char *str);

    /**
     * Flush sendBuffer, i.e., forcefully write its contents to the UART RX
     * characteristic even if the buffer is not full.
     */
    void flush();

    /**
     * Override for Stream::_putc().
     * @param  c
     *         This function writes the character c, cast to an unsigned char, to stream.
     * @return
     *     The character written as an unsigned char cast to an int or EOF on error.
     */
    int _putc(int c);

    /**
     * Override for Stream::_getc().
     * @return
     *     The character read.
     */
    int _getc();
    uint8_t* getRecievedBuffer();

    /**
     * This callback allows the UART service to receive updates to the
     * txCharacteristic. The application should forward the call to this
     * function from the global onDataWritten() callback handler; if that's
     * not used, this method can be used as a callback directly.
     */
    void onDataWritten(const GattWriteCallbackParams *params);
    /*
    This callback allows the connecting device to fetch the endpoint of the device so that it can proceed further.
    */
    void onDataRead(const GattReadCallbackParams *params);


private:
    BLE                &ble;

    uint8_t             receiveBuffer[BLE_UART_SERVICE_MAX_DATA_LEN]; /**< The local buffer into which we receive
                                                                       *   inbound data before forwarding it to the
                                                                       *   application.     */

    uint8_t             sendBuffer[BLE_UART_SERVICE_MAX_DATA_LEN];    /**< The local buffer into which outbound data is
                                                                       *   accumulated before being pushed to the
                                                                       *   rxCharacteristic. */
    char*             _endpointBuffer;                                /*The local buffer that contains the endpoint of
                                                                         the device.       */
    uint8_t              sendBufferIndex;
    uint8_t              numBytesReceived;
    uint8_t              receiveBufferIndex;

    SDAInterface*       _sda_interface;   /*Interface that communicates with SDA */

    GattCharacteristic  txCharacteristic; /**< From the point of view of the external client, this is the characteristic
                                           *   they'd write into in order to communicate with this application. */
    GattCharacteristic  rxCharacteristic; /**< From the point of view of the external client, this is the characteristic
                                           *   they'd read from in order to receive the bytes transmitted by this
                                           *   application. */
    ProtocolTranslator* _protocoltranslator;
};
#endif

